lib/eigensolve_quda.cpp

```diff
@@ -191,6 +191,42 @@ namespace quda
     delete tmp2;
   }
 
+  double EigenSolver::estimateChebyOpMax(const DiracMatrix &mat, ColorSpinorField &out, ColorSpinorField &in)
+  {
+
+    if (in.Location() == QUDA_CPU_FIELD_LOCATION) {
+      in.Source(QUDA_RANDOM_SOURCE);
+    } else {
+      RNG *rng = new RNG(in, 1234);
+      rng->Init();
+      spinorNoise(in, *rng, QUDA_NOISE_UNIFORM);
+      rng->Release();
+      delete rng;
+    }
+
+    ColorSpinorField *in_ptr = &in;
+    ColorSpinorField *out_ptr = &out;
+
+    // Power iteration
+    double norm = 0.0;
+    for (int i = 0; i < 100; i++) {
+      if ((i + 1) % 10 == 0) {
+        norm = sqrt(blas::norm2(*in_ptr));
+        blas::ax(1.0 / norm, *in_ptr);
+      }
+      matVec(mat, *out_ptr, *in_ptr);
+      std::swap(out_ptr, in_ptr);
+    }
+
+    // Compute spectral radius estimate
+    double result = blas::reDotProduct(*out_ptr, *in_ptr);
+    // Zero out workspace
+    blas::zero(out);
```

Comment:
Do we need these two calls to `blas::zero`, i.e., does a subsequent part of the algorithm assume one or both of these are zero? Otherwise we can save the cost of the call.

Reply:
The Krylov space passed to the eigensolver from the C interface is zeroed out. I think it's asking for trouble to populate two vectors with essentially junk in a minor function like this, then handing them back to the eigenslolver. Given that zeroing out two vectors, one time, will not have any significant impact on performance, I'd prefer it if we maintained the zeroed out state of the initial Krylov space.

Reply:
I see it the other way: if the Krylov space vectors don't need to be zeroed out, we shouldn't be zeroing that out either.

I'd let this go if they weren't trivial changes, but since they are trivial, I'd like to see them implemented.

Reply:
I agree with @weinbe2 here.  If there is subsequent zeroing that needs to be done, this should be handled outside of there.  A given function shouldn't be zeroing data proactively just in case for some future use.

Reply:
Ok, took out the zeroing. I had to move the function and use different vectors because as it stood it overwrote potential initial guess data. 

While we're here, I noticed that there is no way for the user to pass in an initial guess via the C interface. Is this intentional? If the device Krylov space referenced the host Krylov space at creation, this would allow for an initial guess. I can hot fix that if there are no objections.